### PR TITLE
Fix  "bind: Address already in use" error

### PR DESCRIPTION
**Pull request by jonnywilliamson**
Leading on from https://github.com/vysheng/tg/issues/431

I had the same issue. If telegram was shut down via CTRL-C (or indeed by a crash) whilst using a tcp port as its socket, any attempt to restart was frequently being met by an error `bind: Address already in use`.

After a certain period of time, the error message would go away (for me 60sec) but i wanted to be able to restart the program immediately.

A look at some online information, showed that this is because the port was left in a TIME_WAIT state and telegram could not attach itself to the port until it was released/timed out.

For some systems the time out can be up to 4 minutes. (Look at the value in `/proc/sys/net/ipv4/tcp_fin_timeout`)
- http://unix.stackexchange.com/questions/72089/how-to-connect-to-port-immedeately-avoiding-socket-bind-unable-to-bind-addre
- http://serverfault.com/questions/329845/how-to-forcibly-close-a-socket-in-time-wait

This PR, uses the `SO_REUSEADDR` option for socket connection which allows the reuse of the port immediately.

### CODE REVIEW

**Comment by luckydonald** [comment]
Somewhere I read, that ```SO_REUSEADDR``` allows the connection from another address - a **new** address.
It still keeps beeing blocked on for the old address (e.g. ```localhost```).

**Comment by jonnywilliamson** [comment]
On my tests here, it allowed the same address to be used again instantly without causing the Address already in use error.

**Comment by jonnywilliamson** [comment]
A VERY detailed description here:
http://stackoverflow.com/questions/14388706/socket-options-so-reuseaddr-and-so-reuseport-how-do-they-differ-do-they-mean-t

TL;DR

>The question is, how does the system treat a socket in state TIME_WAIT? If SO_REUSEADDR is not set, a socket in state TIME_WAIT is considered to still be bound to the source address and port and any attempt to bind a new socket to the same address and port will fail until the socket has really been closed, which may take as long as the configured Linger Time. So don't expect that you can rebind the source address of a socket immediately after closing it. In most cases this will fail. However, if SO_REUSEADDR is set for the socket you are trying to bind, another socket bound to the same address and port in state TIME_WAIT is simply ignored, after all its already "half dead", and your socket can bind to exactly the same address without any problem.

**Comment by jonnywilliamson** [comment]
Sweet!

<runs off and compiles again from source!>

**Comment by luckydonald** [comment]
'Fixes' https://github.com/luckydonald/pytg/issues/21

**Comment by afshinizadi** [comment]
In my case deleting the socket file that was in ` /tmp/tg.sck ` , solved the issue
